pkg/networkpolicy/controller.go

```diff
@@ -656,6 +656,16 @@ func (c *Controller) syncNFTablesRules(ctx context.Context) error {
 		tx.Flush(&knftables.Chain{
 			Name: chainName,
 		})
+
+		if hook == knftables.OutputHook {
+			// accept all outbound traffic to the loopback interface
+			tx.Add(&knftables.Rule{
+				Chain: chainName,
+				Rule: knftables.Concat(
+					"oif", "lo", "accept"),
+			})
```

Comment:
@uablrek can you check in your PR this solves the problem?

Reply:
I added this update, but the address matches should never hit anything with "lo" since such packet would be a "martian"

Reply:
It worked! My world is crumbling! My reasoning is that since the set's only contains pod-addresses, packets matching whose sets can't be routed to/from "lo", since that would make them martians.

Unless... a SatefulSet pod actually have a loopback address as pod-address.

Reply:
@astoycos @tssurya do you have an explanation for this failurs on the conformance suite only with ipv6?

Reply:
The modified chain for reference:
```
        chain output {
                type filter hook output priority filter - 5; policy accept;
                ct state established,related accept
                oif "lo" accept
                ip saddr @podips-v4 queue to 100 comment "process IPv4 traffic with network policy enforcement"
                ip daddr @podips-v4 queue to 100 comment "process IPv4 traffic with network policy enforcement"
                ip6 saddr @podips-v6 queue to 100 comment "process IPv6 traffic with network policy enforcement"
                ip6 daddr @podips-v6 queue to 100 comment "process IPv6 traffic with network policy enforcement"
        }
```

Reply:
it is not a flake, it is a test failure

Reply:
This rule definitely seems wrong.

I think sometimes packets are considered by the kernel to be traversing `lo` if they have the same local source and destination IP, even if that IP isn't 127.0.0.1/::1?

Anyway, the failure here is probably caused by the fact that in ipvs, packets traverse the network stack twice (from source to `ipvs0` and then from `ipvs0` to destination), and the DNAT and SNAT don't happen in exactly the same place as with iptables/nftables kube-proxy. (The DNAT happens inside `ipvs0`... I forget if the SNAT happens in the first or second traversal.)

Anyway, point is, the iptables/nftables logic is going to end up doing the wrong thing for some packets in ipvs mode. The fix _might_ be to ignore packets with `oif = "ipvs0"` so you only look at them on the second traversal, but I'm really not sure about that.

Reply:
this rule fixes the IPv6 jobs, ipvs is unrelated here

Reply:
> I think sometimes packets are considered by the kernel to be traversing lo if they have the same local source and destination IP, even if that IP isn't 127.0.0.1/::1?

So, what I want to express is "we only need to process in OUTPUT local generated packets that are sent outside (not localhost dest)"

![image](https://github.com/kubernetes-sigs/kube-network-policies/assets/6450081/9793fa29-8e90-4cf3-bcf8-2a33fa5b2225)

I had this issue in the past with kubelet probes and we solved it using the lo rule, I think it is correct but I can not explain the science behind it, just the evidence it works as expected

Reply:
> this rule fixes the IPv6 jobs, ipvs is unrelated here

oh... so... the issue is that if we add processing on the `output` hook, then IPv6 (but not IPv4) breaks unless we also add the `! lo` rule?

> So, what I want to express is "we only need to process in OUTPUT local generated packets that are sent outside (not localhost dest)"

Why should it matter though? We shouldn't be blocking any packets that are destined for localhost. With the `podips` changes, we shouldn't even intercept them...

> I can not explain the science behind it

boo!

> any packet destined to localhost does not have to be processed by network policies anyway, right?

A packet from a pod-network pod addressed to `127.0.0.0/8` or `::1` would not leave the pod netns, and it's not possible (by validation) to have an EndpointSlice that points to `127.0.0.0/8` or `::1`, so you can't end up with that via Service DNAT either. But you can add non-localhost IPs to `lo` I think?